proc-macros/src/lib.rs

```diff
@@ -111,212 +110,278 @@ pub fn rpc_client_api(input_token_stream: TokenStream) -> TokenStream {
+///
+/// - This trait will have one additional (already implemented) method `into_rpc`, which
+///   will turn any object that implements the server trait into an `RpcModule`.
+/// - For subscription methods, there will be one additional argument inserted right
+///   after `&self`: `subscription_sink: SubscriptionSink`. It should be used to
+///   actually maintain the subscription.
+///
+/// Since this macro can generate up to two traits, both server and client traits will have
+/// a new name. For the `Foo` trait, server trait will be named `FooServer`, and client,
+/// correspondingly, `FooClient`.
+///
+/// `FooClient` in that case will only have to be imported in the context and will be ready to
+/// use, while `FooServer` must be implemented for some type first.
+///
+/// ## Prerequisites
+///
+/// - Implementors of the server trait must be `Sync`, `Send`, `Sized` and `'static`.
+///   If you want to implement this trait to some type that is not thread-safe, consider
+///   using `Arc<RwLock<..>>`.
+///
+/// ## Examples
+///
+/// Below you can find the example of the macro usage along with the code
+/// that will be generated for it.
+///
+/// ```ignore
+/// #[rpc(client, server, namespace = "foo")]
+/// pub trait Rpc {
+///     #[method(name = "foo")]
+///     async fn async_method(&self, param_a: u8, param_b: String) -> u16;
+///     #[method(name = "bar")]
+///     fn sync_method(&self) -> String;
+///     
+///     #[subscription(name = "sub", unsub = "unsub", item = "String")]
+///     fn sub(&self);
+/// }
+/// ```
+///
+/// Server code that will be generated:
+///
+/// ```ignore
+/// #[async_trait]
+/// pub trait RpcServer {
+///     // RPC methods are usual methods and can be either sync or async.
+///     async fn async_method(&self, param_a: u8, param_b: String) -> u16;
+///     fn sync_method(&self) -> String;
+///
+///     // Note that `subscription_sink` was added automatically.
+///     fn sub(&self, subscription_sink: SubscriptionSink);
+///
+///     fn into_rpc(self) -> Result<Self, jsonrpsee::types::error::Error> {
+///         // Actual implementation stripped, but inside we will create
+///         // a module with one method and one subscription
+///     }
+/// }
+/// ```
+///
+/// Client code that will be generated:
+///
+/// ```ignore
+/// #[async_trait]
+/// pub trait RpcClient: SubscriptionClient {
+///     // In client implementation all the methods are (obviously) async.
+///     async fn async_method(&self, param_a: u8, param_b: String) -> Result<u16, Error> {
+///         // Actual implementations are stripped, but inside a corresponding `Client` or
+///         // `SubscriptionClient` method is called.
+///     }
+///     async fn sync_method(&self) -> Result<String, Error> {
+///         // ...
+///     }
+///     
+///     // Subscription method returns `Subscription` object in case of success.
+///     async fn sub(&self, subscription_sink: SubscriptionSink) -> Result<Subscription<String>, Error> {
```

Comment:
right, I don't understand `subscription_sink: SubscriptionSink` that should be for the server?

It should work to do `async fn sub_with_input(&self, a: usize, b: usize) -> Result<Subscription<String>, Error>` right?

Basically to tell the server to use params as input for the subscription.

Reply:
Great catch! Yeah, the sink is for the server only.